### ISSUE TITLE
Download a file process by server, with HtmlUnit

### ISSUE THREAD

**Issue by Aldarme** [HtmlUnit/htmlunit#352]
Hi every one:
as ask in this stackoverflow topic: [https://stackoverflow.com/questions/66573770/download-a-file-process-by-server-with-htmlunit](url), I open this "issue" to find a way to solve my problem with HTMLUnit.
Thank a lot for all comments,

---------

I want to automatize file conversion available at: 
[https://www.gpsvisualizer.com/map_input?form=googleearth](url)
My problem is that, gpsvisualizer allow standalone conversion, but I have 500 files to convert. So I used hmtlUnit to automatize the process.

Thank to the following code, I am able to modify "select" such as:

    "Output file type"
    "Add DEM elevation data"

upload my file and get the url of the redirected html page where I can download the wanted file.

My problem, is that I do not find a way to download the file.

Does any one have suggestion ?

Thank, in advance.

Here is my code:
`
WebClient webClient = new WebClient();
    webClient.getOptions().setCssEnabled(false);
    webClient.getOptions().setJavaScriptEnabled(true);
    webClient.getOptions().setThrowExceptionOnFailingStatusCode(false);
    webClient.getOptions().setThrowExceptionOnScriptError(false);
    webClient.getOptions().setRedirectEnabled(true);
    
    

    //fetching the web page
    String url = "https://www.gpsvisualizer.com/map_input?form=googleearth";
    //String url = "https://www.reddit.com/r/scraping/";
    HtmlPage page = webClient.getPage(url);
    
    System.out.println(page.getUrl());
    
    System.out.println(page.getTitleText());
    
    //Select set .kml file
    HtmlSelect selectFileType = (HtmlSelect) page.getElementByName("googleearth_zip");
    System.out.println(selectFileType.getOption(0).asText());
    //System.out.println(selectFileType.getOption(1).asText());
    
    HtmlOption kmlFile = selectFileType.getOptionByText(".kml (uncompressed)");
    System.out.println(kmlFile.asText());
    selectFileType.setSelectedAttribute(kmlFile, true);
    
    //Select add elevation on file
    HtmlSelect selectelevation = (HtmlSelect) page.getElementByName("add_elevation");
    System.out.println(selectelevation.getOption(4).asText());
    
    HtmlOption europeSRTM1 = selectelevation.getOptionByText("NASA SRTM1 (30m res., NoAm, Europe, more)");
    System.out.println(europeSRTM1.asText());
    selectelevation.setSelectedAttribute(europeSRTM1, true);
    
    //add file
    HtmlForm myForm = page.getFormByName("main");
    HtmlFileInput fileInput = myForm.getInputByName("uploaded_file_1");
    fileInput.setValueAttribute("/media/Stock/Projets/Suratram/Ressources/Traces_WS/puissance/kml_files/01_douce-signoret.kml");
    HtmlElement submitBtn = page.getElementByName("submitted");
    
    //page google
    HtmlPage page2 = submitBtn.click();
    System.out.println(page2.getUrl());

`

**Comment by rbri** [comment]
As first step you can have a look at https://htmlunit.sourceforge.io/filedownload-howto.html.

Please give me a sign if this does not help.

**Comment by Aldarme** [comment]
I will try this,
and I will give you the result.
Thanks, a lot.

**Comment by Aldarme** [comment]
Hi,
So following your recommendation,
I do my work using the given link, but a have a problem.

After having "click" to allow the server to process my file,
I am move to an other page, and I success to redirect to this new page:
`
page = submitBtn.click();
`
After that, according to the doc., I have to click the 'anchor/button that forces a file download'.

My problem is that, the anchor is identify as follow:
`<a href="/display/20210513073349-38053-map.kmz">20210513073349-38053-map.kmz</a>`

with no Name, Id, or anything else,

and this part: "20210513073349-38053-map.kmz", change each time I reach the new page.

So, I tried to list all anchors on the page:
`
List<HtmlAnchor> anchors = new ArrayList<HtmlAnchor>();
	    
	    anchors = page.getAnchors();
	    
	    for(HtmlAnchor elm : anchors) {
	    	System.out.println(elm);
	    }
`

But the wanted anchor, above mention, is never present in the list.

So I don't know how to processed to continue.

I am maybe wrong about the way of doing this,
Does it possible to help  me or provide me advice to find a way ?

Best regard,
Aldarme

**Comment by Aldarme** [comment]
Hi,
After having work with resources of your previous link, I come back do to a problem.

After having follow the redirection due to the click to allow the server to process my file
and "click some anchor/button that forces a file download"
`  

      //Arrive on the redirected page, where to download the file
	  page = submitBtn.click();
	  //System.out.println(page.getUrl());
	  
	  WebWindow window = page.getEnclosingWindow();
	  
	  //System.out.println( page.asNormalizedText() );
	  
	  // click the anchor that forces the file download
	  for(HtmlAnchor elm : page.getAnchors()) {
	   //System.out.println(elm);
	   
	   if(elm.asNormalizedText().contains("-map.kml")) {
	    System.out.println(elm);
	    elm.click();
	   }

`

So after, I continue with the download of my file:
`

      UnexpectedPage downloadPage = (UnexpectedPage) window.getEnclosedPage();
	    
      InputStream downloadedContent = downloadPage.getInputStream();
    	
      File file = new File("PATH/test.kml");
    	
      OutputStream output = new FileOutputStream(file, false);
    	
      downloadedContent.transferTo(output);

`

I have a problem with the cast type of the "WebWindow" as "UnexpectedPage".

The console display:
`
class com.gargoylesoftware.htmlunit.TopLevelWindow cannot be cast to class com.gargoylesoftware.htmlunit.UnexpectedPage
`
I do not understand because the documentation specify that " If the response type is supported the content will be available as HtmlPage (or XHtmlPage)", so I do not understand why this trubble.
Furthermore, I do not find working solution on stackOverflow or other.

I should miss something.

Can you please, give me some tips ?

Best,
Aldarme

**Comment by rbri** [comment]
> class com.gargoylesoftware.htmlunit.TopLevelWindow cannot be cast to class com.gargoylesoftware.htmlunit.UnexpectedPage

The page is inside the window - window.getEnclosedPage()

**Comment by Aldarme** [comment]
Ok, but I don't understand your answer.
I include my new page on the webWindow

`

     //Arrive on the redirected page, where to download the file
     page = submitBtn.click();  
     WebWindow window = page.getEnclosingWindow();

`
and use the "window.getEnclosedPage()" as specified in the doc.
`

       UnexpectedPage downloadPage = (UnexpectedPage) window.getEnclosedPage();
	    
	InputStream downloadedContent = downloadPage.getInputStream();
    	
    	File file = new File("/home/promet/Bureau/test/test.kml");
    	
    	OutputStream output = new FileOutputStream(file, false);
    	
    	downloadedContent.transferTo(output);

`

Can you provide me more explanation, please ?

**Comment by rbri** [comment]
can you please provide a sample file and i will try to build a working sample for you

**Comment by Aldarme** [comment]
Here is my code

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import org.w3c.dom.Node;

import com.gargoylesoftware.htmlunit.*;
import com.gargoylesoftware.htmlunit.html.*;
import com.gargoylesoftware.htmlunit.html.parser.HTMLParser;

public class AltScrapper {

	public static void main(String[] args) throws Exception {
		
		WebClient webClient = new WebClient();
		webClient.getOptions().setUseInsecureSSL(true);
	    webClient.getOptions().setCssEnabled(false);
	    webClient.getOptions().setJavaScriptEnabled(false);	    

	    //fetching the web page
	    String url = "https://www.gpsvisualizer.com/map_input?form=googleearth";
	    HtmlPage page = webClient.getPage(url);
	    
	    System.out.println(page.getUrl());
	    
	    System.out.println(page.getTitleText());
	    
	    //Select set .kml file
	    HtmlSelect selectFileType = (HtmlSelect) page.getElementByName("googleearth_zip");
	    System.out.println(selectFileType.getOption(0).asNormalizedText());
	    //System.out.println(selectFileType.getOption(1).asText());
	    
	    HtmlOption kmlFile = selectFileType.getOptionByText(".kml (uncompressed)");
	    System.out.println(kmlFile.asNormalizedText());
	    selectFileType.setSelectedAttribute(kmlFile, true);
	    
	    //Select add elevation on file
	    HtmlSelect selectelevation = (HtmlSelect) page.getElementByName("add_elevation");
	    System.out.println(selectelevation.getOption(4).asNormalizedText());
	    
	    HtmlOption europeSRTM1 = selectelevation.getOptionByText("NASA SRTM1 (30m res., NoAm, Europe, more)");
	    System.out.println(europeSRTM1.asNormalizedText());
	    selectelevation.setSelectedAttribute(europeSRTM1, true);
	    
	    //add file
	    HtmlForm myForm = page.getFormByName("main");
	    HtmlFileInput fileInput = myForm.getInputByName("uploaded_file_1");
	    fileInput.setValueAttribute("/media/Stock/Projets/Suratram/Ressources/Traces_WS/GoogleEarth/ligne1/bus9/kml/segments_7-8/finalFiles/1_LaDouce-Signoret.kml");
	    HtmlElement submitBtn = page.getElementByName("submitted");
	    
	    //Arrive on the redirected page, where to download the file
	    page = submitBtn.click();
	    //System.out.println(page.getUrl());
	    
	    WebWindow window = page.getEnclosingWindow();
	    
	    //System.out.println( page.asNormalizedText() );
	    
	    // click the anchor that forces the file download
	    for(HtmlAnchor elm : page.getAnchors()) {
	    	//System.out.println(elm);
	    	
	    	if(elm.asNormalizedText().contains("-map.kml")) {
	    		System.out.println(elm);
	    		elm.click();
	    	}	    	
	    }
	    
	    UnexpectedPage downloadPage = (UnexpectedPage) window.getEnclosedPage();
	    
	    InputStream downloadedContent = downloadPage.getInputStream();
    	
    	File file = new File("/home/promet/Bureau/test/test.kml");
    	
    	OutputStream output = new FileOutputStream(file, false);
    	
    	downloadedContent.transferTo(output);
		
		
	}

}

**Comment by Aldarme** [comment]
here is a exemple file, I need to convert:

<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
<Document>
	<name>douce-signoret.kml</name>
	<StyleMap id="1_green">
		<Pair>
			<key>normal</key>
			<styleUrl>#1_green1</styleUrl>
		</Pair>
		<Pair>
			<key>highlight</key>
			<styleUrl>#1_green0</styleUrl>
		</Pair>
	</StyleMap>
	<Style id="1_green0">
		<IconStyle>
			<Icon>
				<href>http://www.rttb.fr/google_earth/1_green.png</href>
			</Icon>
			<hotSpot x="0" y="0" xunits="pixels" yunits="pixels"/>
		</IconStyle>
	</Style>
	<Style id="1_green1">
		<IconStyle>
			<Icon>
				<href>http://www.rttb.fr/google_earth/1_green.png</href>
			</Icon>
			<hotSpot x="0" y="0" xunits="pixels" yunits="pixels"/>
		</IconStyle>
	</Style>
	<Style id="1_green2">
		<IconStyle>
			<Icon>
				<href>http://www.rttb.fr/google_earth/1_green.png</href>
			</Icon>
			<hotSpot x="0" y="0" xunits="pixels" yunits="pixels"/>
		</IconStyle>
	</Style>
	<Folder>
		<name>douce-signoret</name>
		<Placemark>
			<name>9 (07:05:11)</name>
			<Snippet maxLines="2"><![CDATA[Date : 30/10/2020<br/>Heure : 07:05:11<br/>Etat : En course]]></Snippet>
			<description><![CDATA[Date : 30/10/2020<br/>Heure : 07:05:11<br/>Etat : En course<br/>Conducteur : RIPAMONTI CHRISTELLE<br/>Ligne : 1<br/>Destination : Résidences<br/>Dernier arret : Dépôt RTTB<br/>Service voiture : 1/7<br/>Service Agent : 1115<br/>Retard : 0 ' 27<br/>Latitude : 47,629<br/>Longitude : 6,8402

   ]]></description>
			<styleUrl>#1_green</styleUrl>
			<Point>
				<coordinates>6.835769444444444,47.62759166666667,0</coordinates>
			</Point>
		</Placemark>
		<Placemark>
			<name>9 (07:07:40)</name>
			<Snippet maxLines="2"><![CDATA[Date : 30/10/2020<br/>Heure : 07:07:40<br/>Etat : En Terminus]]></Snippet>
			<description><![CDATA[Date : 30/10/2020<br/>Heure : 07:07:40<br/>Etat : En Terminus<br/>Conducteur : RIPAMONTI CHRISTELLE<br/>Ligne : 1<br/>Destination : Valdoie Mairie<br/>Dernier arret : La Douce<br/>Service voiture : 1/7<br/>Service Agent : 1115<br/>Retard : 0 ' 40<br/>Latitude : 47,6278<br/>Longitude : 6,8374]]></description>
			<Point>
				<coordinates>6.837400000000001,47.6278,0</coordinates>
			</Point>
		</Placemark>
		<Placemark>
			<name>9 (07:07:44)</name>
			<Snippet maxLines="2"><![CDATA[Date : 30/10/2020<br/>Heure : 07:07:44<br/>Etat : En course]]></Snippet>
			<description><![CDATA[Date : 30/10/2020<br/>Heure : 07:07:44<br/>Etat : En course<br/>Conducteur : RIPAMONTI CHRISTELLE<br/>Ligne : 1<br/>Destination : Valdoie Mairie<br/>Dernier arret : La Douce<br/>Service voiture : 1/7<br/>Service Agent : 1115<br/>Retard : 0 ' 15<br/>Latitude : 47,6279<br/>Longitude : 6,8377]]></description>
			<styleUrl>#1_green2</styleUrl>
			<Point>
				<coordinates>6.8377,47.6279,0</coordinates>
			</Point>
		</Placemark>
		<Placemark>
			<name>9 (07:08:15)</name>
			<Snippet maxLines="2"><![CDATA[Date : 30/10/2020<br/>Heure : 07:08:15<br/>Etat : En course]]></Snippet>
			<description><![CDATA[Date : 30/10/2020<br/>Heure : 07:08:15<br/>Etat : En course<br/>Conducteur : RIPAMONTI CHRISTELLE<br/>Ligne : 1<br/>Destination : Valdoie Mairie<br/>Dernier arret : La Douce<br/>Service voiture : 1/7<br/>Service Agent : 1115<br/>Retard : 0 ' 08<br/>Latitude : 47,6291<br/>Longitude : 6,8407]]></description>
			<styleUrl>#1_green2</styleUrl>
			<Point>
				<coordinates>6.840699999999999,47.62909999999999,0</coordinates>
			</Point>
		</Placemark>
		<Placemark>
			<name>9 (07:08:20)</name>
			<Snippet maxLines="2"><![CDATA[Date : 30/10/2020<br/>Heure : 07:08:20<br/>Etat : En course]]></Snippet>
			<description><![CDATA[Date : 30/10/2020<br/>Heure : 07:08:20<br/>Etat : En course<br/>Conducteur : RIPAMONTI CHRISTELLE<br/>Ligne : 1<br/>Destination : Valdoie Mairie<br/>Dernier arret : Signoret<br/>Service voiture : 1/7<br/>Service Agent : 1115<br/>Retard : 0 ' 10<br/>Latitude : 47,6295<br/>Longitude : 6,8406]]></description>
			<styleUrl>#1_green2</styleUrl>
			<Point>
				<coordinates>6.8406,47.6295,0</coordinates>
			</Point>
		</Placemark>
	</Folder>
</Document>
</kml>

**Comment by Aldarme** [comment]
Here is a sample file to convert.
Be carefull, I had to add ".txt" extension because, git do not let me upload ".KML" file,
so remove ".txt" extention.

[1_LaDouce-Signoret.kml.txt](https://github.com/HtmlUnit/htmlunit/files/6496140/1_LaDouce-Signoret.kml.txt)

Best,
Aldarme

**Comment by rbri** [comment]
Here it is - this should work with zipped and unzipped responses.
For unzipped the server returns an Xml file - in this case the page an XmlPage not an UnexpectedPage - sorry.
(this uses apache commons io)

        try (WebClient webClient = new WebClient(BrowserVersion.FIREFOX)) {
            webClient.getOptions().setThrowExceptionOnScriptError(false);

            final HtmlPage page = webClient.getPage("https://www.gpsvisualizer.com/map_input?form=googleearth");
            webClient.waitForBackgroundJavaScriptStartingBefore(10_000);

            // Select set .kml file
            HtmlSelect selectFileType = (HtmlSelect) page.getElementByName("googleearth_zip");
            // HtmlOption kmlFile = selectFileType.getOptionByText(".kml (uncompressed)");
            HtmlOption kmlFile = selectFileType.getOptionByText(".kmz (zipped)");
            kmlFile.setSelected(true);

            // Select add elevation on file
            HtmlSelect selectelevation = (HtmlSelect) page.getElementByName("add_elevation");
            HtmlOption europeSRTM1 = selectelevation.getOptionByText("NASA SRTM1 (30m res., NoAm, Europe, more)");
            europeSRTM1.setSelected(true);

            // add file
            HtmlForm myForm = page.getFormByName("main");
            HtmlFileInput fileInput = myForm.getInputByName("uploaded_file_1");
            fileInput.setFiles(new File("C:\\RBRi\\eclipse\\workspace\\HtmlUnit Support\\src\\issue352\\1_LaDouce-Signoret.kml"));

            HtmlElement submitBtn = page.getElementByName("submitted");
            HtmlPage page2 = submitBtn.click();

            // search for the download anchor
            for (final HtmlAnchor anchor : page2.getAnchors()) {
                // support kmk / kmz
                if (anchor.asNormalizedText().contains("-map.km")) {
                    Page page3 = anchor.click();

                    try (InputStream is = page3.getWebResponse().getContentAsStream()) {
                        try (OutputStream os = new FileOutputStream(new File("C:\\RBRi\\eclipse\\workspace\\HtmlUnit Support\\src\\issue352\\result"))) {
                            IOUtils.copy(is, os);
                        }
                    }
                }
            }
        }

Hope that helps.

**Comment by Aldarme** [comment]
Hi rbri,
Thanks a lot, I will try this and give you update.
Best,
Aldarme

**Comment by Aldarme** [comment]
Hi,

I had some issues with a part of your code,
so I update mine and It's working.

Thank a lot for the time spent with helping me.

Best regards,
Aldarme

**Comment by rbri** [comment]
Your welcome, have fun using HtmlUnit and please update the stackoverflow question
